tests/acceptance/dex/idtokenprovider.go

```diff
@@ -0,0 +1,116 @@
+package dex
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"log"
+
+	"github.com/pkg/errors"
+)
+
+type idTokenProvider interface {
+	fetchIdToken() (string, error)
+}
+
+type dexIdTokenProvider struct {
+	httpClient *http.Client
+	config     idProviderConfig
+}
+
+func newDexIdTokenProvider(httpClient *http.Client, config idProviderConfig) idTokenProvider {
+	return &dexIdTokenProvider{httpClient, config}
+}
+
+func (p *dexIdTokenProvider) fetchIdToken() (string, error) {
+
+	flowResult, err := p.implicitFlow()
+	if err != nil {
+		return "", err
+	}
+	return flowResult["id_token"], nil
+}
+
+func (p *dexIdTokenProvider) implicitFlow() (map[string]string, error) {
+
+	authorizeResp, err1 := p.httpClient.PostForm(p.config.dexConfig.authorizeEndpoint, url.Values{
+		"response_type": {"id_token token"},
+		"client_id":     {p.config.clientConfig.id},
+		"redirect_uri":  {p.config.clientConfig.redirectUri},
+		"scope":         {"openid profile email groups"},
+		"nonce":         {"vF7FAQlqq41CObeUFYY0ggv1qEELvfHaXQ0ER4XM"},
+	})
+	if err1 != nil {
+		return nil, err1
+	}
+	if authorizeResp.StatusCode < 300 || authorizeResp.StatusCode > 399 {
+		return nil, errors.New(fmt.Sprintf("Authorize - response error: '%s' - %s", authorizeResp.Status, readRespBody(authorizeResp)))
+	}
+
+	// /auth/local?req=qruhpy2cqjvv4hcrbuu44mf4v
+	loginEndpoint := authorizeResp.Header.Get("location")
+	if strings.Contains(loginEndpoint, "#.*error") {
+		return nil, errors.New(fmt.Sprintf("Login - Redirected with error: '%s'", loginEndpoint))
+	}
+
+	_, err2 := p.httpClient.Get(p.config.dexConfig.baseUrl + loginEndpoint)
+	if err2 != nil {
+		return nil, err2
+	}
+
+	loginResp, err3 := p.httpClient.PostForm(p.config.dexConfig.baseUrl+loginEndpoint, url.Values{
+		"login":    {p.config.userCredentials.username},
+		"password": {p.config.userCredentials.password},
+	})
+	if err3 != nil {
+		return nil, err3
+	}
+	if loginResp.StatusCode < 300 || loginResp.StatusCode > 399 {
+		return nil, errors.New(fmt.Sprintf("Login - response error: '%s' - %s", loginResp.Status, readRespBody(loginResp)))
```

Comment:
Use `fmt.Errorf` in all places, where you use `errors.New(fmt.Sprintf(...))`

Reply:
done